Competitive_Programming/C++/DP/Edit Distance/edit_distance_documentation.md

```diff
@@ -0,0 +1,149 @@
+Example: 
+``` 
+Input: str1="sitting",  str2="kitten"
+Output: 3
+Explanation: 2 Replace and 1 Remove operation can be performed, i.e, replace character 's' with 'k' and character 'i' with 'e' and remove last character 'g'. 
+
+Input: str1="girl",  str2="girlscript"
+Output: 6
+Explanation: insert 6 characters after str1 i.e 's', 'c', 'r', 'i', 'p', 't'
+```
+
+## Naive Approach 🧠
+
+A naive brute force idea is to traverse on the strings and process each character one by one. Let traverse both the strings from right to left. 
+m = Length of String 1 
+n = Lenght of String 2
+There can be two possiblity 
+
+**Case 1**: Last Characters of both the Strings are same. Nothing to Do, so cost=0 move to next index, i.e, Recur for (n-1,m-1)
+
+**Case 2**: Last Characters of both the Strings are different. Now we have 3 Operations that we can perform either, 
+ 1. Insert the same character at that index i.e, recur for (n,m-1)
+ 2. Remove that character at that index, i.e recur for (n-1,m)
+ 3. Replace with same character at that index i.e recur for (n-1,m-1)
+ and add cost 1 to the answer. 
+Take the minimum of all the operations. 
+
+## 💡 Implementation of Naive Approach: 
+```c++
+int edit_distance_recursive(string str1, string str2, int n, int m)
+{
+    // If string 2 is empty then remove all character of string 1
+    if(m==0) return n;  
+    // If string 1 is empty then insert all character of string 2
+    if(n==0) return m;  
+
+    // If at ith index both are same no operations needed to be performed 
+    if(str1[n-1]==str2[m-1])  {
+        return edit_distance_recursive(str1,str2,n-1,m-1); 
+    }
+    // Consider all possible Operations 
+    int insert = edit_distance_recursive(str1,str2,n,m-1); 
+    int remove = edit_distance_recursive(str1,str2,n-1,m);
+    int replace = edit_distance_recursive(str1,str2,n-1,m-1); 
+
+    // Return minimum of all operations 
+    return 1+min({insert,remove,replace});         
+}
+```
+
+#### ⏰ Time Complexity for Naive Approach 
+
+__Time Complexity for the Naive Approach is Exponential__. Since we are performing 3 Recursion Calls so in worst Case we may end up with **O(3<sup>n</sup>)** operations. 
+
+![image](https://user-images.githubusercontent.com/58984074/134164390-56828bac-3a4f-4fbd-9b9b-3211fdf02dd1.png)
+
+It can be concluded that are lots of overlapping subproblems that are solved again and again. Example f(2,2) is calculated many times. Hence we can store the intermediate results of computations and use them further if required. 
+Hence a **_Dynamic Programming_** Recursion + Memoization approach can be used here.  
+
+
+## 💡 Implementation of Recursion + Memoization (Top Down Approach):
+
+The above Naive Recursive approach above can be optimized using memoization. A 2-D Vector can be used to store the intermediate results and for every state f(n,m) it can be checked whether it has been calculated before. 
```

Comment:
```suggestion
The above Naive Recursive approach above can be optimized using memoization. A 2-D vector can be used to store the intermediate results and for every state f(n,m) it can be checked whether it has been calculated before. 
```